check/bench.go

```diff
@@ -0,0 +1,128 @@
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+package check
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/golang/glog"
+	"gopkg.in/yaml.v2"
+)
+
+type Bench interface {
+	RegisterAuditType(auditType AuditType, typeCallback func() interface{}) error
+	NewControls(in []byte, definitions []string, customConfigs ...interface{}) (*Controls, error)
+}
+
+type bench struct {
+	auditTypeRegistry map[AuditType]func() interface{}
+}
+
+func NewBench() Bench {
+	return &bench{auditTypeRegistry:make(map[AuditType]func() interface{})}
+}
+
+func (b *bench) RegisterAuditType(auditType AuditType, typeCallback func() interface{}) error {
+
+	if _, ok := b.auditTypeRegistry[auditType]; ok {
+		return fmt.Errorf("audit type %v already registered", auditType)
+	}
+	a := typeCallback()
+	if _, ok := a.(Auditer); ok {
+		b.auditTypeRegistry[auditType] = typeCallback
+		return nil
+	}
+	return fmt.Errorf("audit type %v must implement Auditer interface", auditType)
+
+}
+
+func (b *bench) NewControls(in []byte, definitions []string, customConfigs ...interface{}) (*Controls, error) {
+	c := new(Controls)
+	//c.auditTypeRegistry = defaultBench.auditTypeRegistry
+	err := yaml.Unmarshal(in, c)
+	if err != nil {
+		return nil, fmt.Errorf("failed to unmarshal YAML: %s", err)
+	}
+	c.customConfigs = customConfigs
+	if len(definitions) > 0 {
+		c.DefinedConstraints = map[string][]string{}
+		for _, val := range definitions {
+			a := strings.Split(val, "=")
+
+			// If its type 'category=value' for example 'platform=ubuntu'
+			if len(a) == 2 && a[0] != "" && a[1] != "" {
+				c.DefinedConstraints[a[0]] = append(c.DefinedConstraints[a[0]], a[1])
+			} else {
+				glog.V(1).Info("failed to parse defined constraint, ", val)
+			}
+		}
+	}
+	if err := b.extractAllAudits(c); err != nil {
+		return nil, err
+	}
+	return c, nil
+}
+
+func (b *bench) convertAuditToRegisteredType(auditType AuditType, audit interface{}) (auditer Auditer, err error) {
+
+	var auditBytes []byte
+	var callback func() interface{}
+	var ok bool
+
+	if auditType == "" || auditType == TypeAudit {
+		if s, ok := audit.(string); ok || audit == nil {
+			return Audit(s), nil
+		}
+		return nil, fmt.Errorf("failed to convert audit, mismatching type")
+	}
+
+	if callback, ok = b.auditTypeRegistry[auditType]; !ok {
+		return nil, fmt.Errorf("audit type %v is not registered", auditType)
+	}
+
+	o := callback()
+	if auditBytes, err = yaml.Marshal(audit); err != nil {
+		return nil, fmt.Errorf("unable to marshal Audit %v", err)
+	}
+
+	if err := yaml.Unmarshal(auditBytes, o); err != nil {
+		return nil, fmt.Errorf("unable to Unmarshal Audit %v", err)
+	}
+	return o.(Auditer), nil
```

Comment:
Would be an idea to check that `o` can be converted to `Auditer`, either here or when it’s created at line 95

Reply:
I don't mind to add such check , however during type registration I am already doing this check, so the  auditTypeRegistry factory always contains structures that implements Auditer, thus adding check here will be redundant  
see line 43 
	if _, ok := a.(Auditer); ok {
		b.auditTypeRegistry[auditType] = typeCallback
		return nil
	}

Reply:
I'll accept it for today. 

I agree it is already checked earlier as the code stands today, but this relies on nothing changing the auditTypeRegistry between it being set up, and it being used here. For future maintainability it would be nice to have a check here as well.